File: src/modules/hardware/lib/sensorReadings.ts
import { theme } from '@/constants/theme'

/**
 * How one frame key is shown. Presentation only: what a raw value means — its scale, its range,
 * whether it is worth a chart — is the firmware contract and is decided natively, so the numbers
 * arriving here are already in display units.
 *
 * @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/hardware/SensorReadings.kt `ReadingSpec`
 */
interface ReadingSpec {
  label: string
  unit: string
  decimals: number
  color: string
  /**
   * Keep this key out of the rows entirely. Link bookkeeping the board sends so the app can
   * measure the connection is reported in the Link section, not as a sensor reading.
   */
  hidden?: true
}

const READINGS: Record<string, ReadingSpec> = {
  distanceMm: {
    label: 'Distance (ToF)',
    unit: 'cm',
    decimals: 1,
    color: theme.palette.cyan.color,
  },
  rangeCm: {
    label: 'Range (ultrasonic)',
    unit: 'cm',
    decimals: 1,
    color: theme.palette.violet.color,
  },
  tempC: {
    label: 'Chip temperature',
    unit: '°C',
    decimals: 1,
    color: theme.palette.amber.color,
  },
  heapKb: {
    label: 'Free heap',
    unit: 'kB',
    decimals: 0,
    color: theme.palette.green.color,
  },
  seq: {
    label: 'Frame',
    unit: '',
    decimals: 0,
    color: theme.neutral.textMuted,
    hidden: true,
  },
  readMs: {
    label: 'Sensor read',
    unit: 'ms',
    decimals: 0,
    color: theme.neutral.textMuted,
    hidden: true,
  },
  upMs: {
    label: 'Uptime',
    unit: 's',
    decimals: 0,
    color: theme.neutral.textMuted,
  },
}

const UNKNOWN: Omit<ReadingSpec, 'label'> = {
  unit: '',
  decimals: 1,
  color: theme.palette.sky.color,
}

export interface Reading {
  key: string
  label: string
  unit: string
  decimals: number
  color: string
  hidden: boolean
}

/** Everything a frame key needs to be rendered, whether or not the app knows the key. */
export function describeReading(key: string): Reading {
  const spec = READINGS[key] ?? { ...UNKNOWN, label: key }
  return {
    key,
    label: spec.label,
    unit: spec.unit,
    decimals: spec.decimals,
    color: spec.color,
    hidden: spec.hidden === true,
  }
}

/**
 * One row per key, in the order given, minus the link bookkeeping the board sends for itself.
 *
 * Rows carry no value: the numbers arrive far faster than React should render, so a row is the
 * label and the shape, and the value behind it is a shared value the UI thread writes. Rows are
 * kept even once a sensor stops answering, since the ToF drops out whenever nothing is in range
 * and letting the row go makes everything below it jump.
 */
export function describeReadings(keys: readonly string[]): Reading[] {
  return keys.map(describeReading).filter((reading) => !reading.hidden)
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/hardware/SensorReadings.kt
package expo.modules.vescapecore.hardware

/**
 * The numeric half of a board reading: what a raw frame value means, and what may be drawn from
 * it. Labels, units and colors are the app's business and stay in JS; scale, range and
 * chartability are the contract with the firmware and are decided here, once, for both the live
 * numbers and the charts.
 *
 * @parity /src/modules/hardware/lib/sensorReadings.ts `READINGS`
 */
internal data class ReadingSpec(
    /** Raw frame units to display units. Both distance sensors are shown in cm. */
    val scale: Double = 1.0,
    /**
     * Display range. Values are clamped into it, the chart axis is fixed to it, and a sensor that
     * read nothing sits at its top: "no target" means "further than the range", not zero.
     */
    val min: Double? = null,
    val max: Double? = null,
    /** Whether this reading is worth a chart row, or is a number to glance at. */
    val chart: Boolean = false,
)

/** Useful reach for both distance sensors on a board. Anything past this is not a reading. */
private const val DISTANCE_MAX_CM = 40.0

private val SPECS = mapOf(
    "distanceMm" to ReadingSpec(scale = 0.1, min = 0.0, max = DISTANCE_MAX_CM, chart = true),
    "rangeCm" to ReadingSpec(min = 0.0, max = DISTANCE_MAX_CM, chart = true),
    "upMs" to ReadingSpec(scale = 0.001),
)

private val UNKNOWN = ReadingSpec()

/**
 * Readings the app knows a board can take, in the order they are shown.
 *
 * A ranged sensor is declared rather than discovered: "nothing in reach" is a reading, and the
 * firmware leaves the key out of the frame when it gets one. Waiting for a first echo before the
 * row exists means the row appears late, and everything under it jumps when it does.
 */
internal val DECLARED_KEYS: List<String> = SPECS.entries.filter { it.value.max != null }.map { it.key }

internal fun readingSpec(key: String): ReadingSpec = SPECS[key] ?: UNKNOWN

/**
 * A raw frame value in display units, clamped to the key's range, or null when the sensor read
 * nothing and has no ceiling to fall back on.
 */
internal fun readingValue(key: String, raw: Double?): Double? {
    val spec = readingSpec(key)
    if (raw == null) return spec.max
    if (!raw.isFinite()) return null
    val converted = raw * spec.scale
    val min = spec.min
    val max = spec.max
    if (min == null || max == null) return converted
    return converted.coerceIn(min, max)
}
